docs/internal/3i.md

```diff
@@ -0,0 +1,141 @@
+
+From 3i’s perspective, a grate is both:
+
+- A consumer of system calls from descendant cages, and
+- A service provider exposing its own callable entry points to the 3i runtime.
+
+### Design Goals
+
+3i’s public API is runtime-agnostic, allowing it to be adapted to multiple backends (e.g., Wasmtime, MPK, or future runtimes).
+
+Runtime-specific mechanisms, such as Wasmtime’s `VMContext` management, are implemented in the corresponding runtime layer (e.g., the Wasmtime adapter). These runtime hooks are exposed to 3i as C-ABI function pointers and registered in 3i’s global table, decoupling runtime details from 3i’s core logic.
+
+### Wasm Implementation
+
+Grate execution relies on a re-entry mechanism that bridges Wasm runtime contexts (e.g., Wasmtime instances) and 3i’s routing layer. 
+
+Each Wasm module in Wasmtime has its own `VMContext`, which encapsulates all runtime state required for that instance’s execution. When 3i performs a cross-module call (e.g., from _cage A_ to a _grate B_), the currently executing context belongs to _A_, but 3i needs to restore the latest runtime state of _B_ before re-entering its Wasm code. 
+
+Because Rust’s lifetime system prevents us from directly sharing or globally storing references to the caller’s runtime structures, we instead store and pass raw pointers to the target module’s `VMContext`. 3i uses a registered callback function pointer and retrived the corresponding runtime context (`VMContext`) from pointer to re-enter the target Wasm instance. This allows 3i to reconstruct the correct execution environment for _B_ at re-entry time, ensuring cross-instance transitions without violating Rust’s ownership and lifetime rules.
+
+During module initialization, Wasmtime captures each instance’s low-level `VMContext`, representing its active execution state (memories, globals, tables, etc.). However, the registration of interposition handlers (`register_handler`) often occurs after initialization. To bridge this timing gap, Wasmtime maintains a runtime-local table:
+
+```rust
+GRATE_FN_WASM_TABLE: HashMap<(u64, u64), Box<GrateFnEntry>>
+```
+
+- Key: `(cageid, tid)` -- cage and its thread identifiers
+- Value: `Box<GrateFnEntry>` -- stable heap-allocated entry containing callback and necessary context pointers
+
+When the user’s Wasm code calls `register_handler`, the entry is extracted and passed to 3i, where it becomes the canonical reference for that `(cageid, tid)` pair.
+
+_Note: `tid = 0` is currently used as a placeholder; multi-thread support will be added in future revisions._
+
+#### Core Data Structures
+
+**`VmCtxWrapper`**
+
+A minimal wrapper around a non-null pointer to a Wasmtime `VMContext`. It allows the context pointer to be passed between Wasmtime and 3i without exposing the raw pointer across the codebase.
```

Comment:
I'm not sure why 3i needs this.  Still confused here.

Reply:
@Yaxuan-w Is this because fork needs access to know what context is it called from?

Also, why do you care if the raw pointer is exposed?